sklearn/metrics/_scorer.py

```diff
@@ -323,6 +323,12 @@ def _score(self, method_caller, clf, X, y, sample_weight=None):
                                              self._score_func.__name__))
                 elif isinstance(y_pred, list):
                     y_pred = np.vstack([p[:, -1] for p in y_pred]).T
+        else:  # multiclass
+            try:
+                y_pred = method_caller(clf, "predict_proba", X)
```

Comment:
should we try ``decision_function`` first? ``_ThresholdScorer`` means that we "Evaluate decision function output for X relative to y_true".

Reply:
I agree, `_ThresholdScorer` should be consistent and always try to use `decision_function` first whatever the type of output: it's should work for any kind model that has un-normalized but continuous class assignment scores.

If ROC AUC with multiclass averaging needs normalized class assignment probabilities instead, we should use this requirement in the defintions of the scorer instances instead:

Currently, scorer.py has:
```python
# Score functions that need decision values
roc_auc_scorer = make_scorer(roc_auc_score, greater_is_better=True,
                             needs_threshold=True)
average_precision_scorer = make_scorer(average_precision_score,
                                       needs_threshold=True)
roc_auc_ovo_scorer = make_scorer(roc_auc_score, needs_threshold=True,
                                 multi_class='ovo')
roc_auc_ovo_weighted_scorer = make_scorer(roc_auc_score, needs_threshold=True,
                                          multi_class='ovo',
                                          average='weighted')
roc_auc_ovr_scorer = make_scorer(roc_auc_score, needs_threshold=True,
                                 multi_class='ovr')
roc_auc_ovr_weighted_scorer = make_scorer(roc_auc_score, needs_threshold=True,
                                          multi_class='ovr',
                                          average='weighted')
```
If needed, we should replace `needs_threshold=True` by `needs_proba=True` but I do not see why this would be the case for ROC AUC.

Reply:
> If needed, we should replace needs_threshold=True by needs_proba=True but I do not see why this would be the case for ROC AUC.

I see, so we should use ``need_proba=True``, we do not consider decision_function in _ProbaScorer.

For roc_auc_scorer, I think we should use _ThresholdScorer becasue we accept the output from decision_function.